File: rsa.java
import java.math.BigInteger;
import java.util.Scanner;

class Main {
    // Function to compute base^expo mod m using BigInteger
    static BigInteger power(BigInteger base, BigInteger expo, BigInteger m) {
        return base.modPow(expo, m);
    }

    // Function to find modular inverse of e modulo phi(n)
    static BigInteger modInverse(BigInteger e, BigInteger phi) {
        return e.modInverse(phi);
    }

    // Function to verify if a number is prime
    static boolean isPrime(BigInteger num) {
        return num.isProbablePrime(100);
    }

    // Function to check if e is valid
    static boolean isValidE(BigInteger e, BigInteger phi) {
        return e.compareTo(BigInteger.ONE) > 0 && 
               e.compareTo(phi) < 0 && 
               e.gcd(phi).equals(BigInteger.ONE);
    }

    // Function to check if d is valid
    static boolean isValidD(BigInteger d, BigInteger e, BigInteger phi) {
        return d.multiply(e).mod(phi).equals(BigInteger.ONE);
    }

    // RSA Key Generation with user input
    static void generateKeys(BigInteger[] keys, Scanner scanner) {
        while (true) {
            try {
                // Get p and q from user
                System.out.print("Enter first prime number (p): ");
                BigInteger p = new BigInteger(scanner.nextLine());
                
                System.out.print("Enter second prime number (q): ");
                BigInteger q = new BigInteger(scanner.nextLine());

                // Verify if numbers are prime
                if (!isPrime(p) || !isPrime(q)) {
                    System.out.println("Error: Both numbers must be prime!");
                    continue;
                }

                BigInteger n = p.multiply(q);
                BigInteger phi = p.subtract(BigInteger.ONE).multiply(q.subtract(BigInteger.ONE));

                // Get e from user
                System.out.println("phi(n) = " + phi);
                System.out.print("Choose e (1 < e < phi(n) and gcd(e, phi(n)) = 1): ");
                BigInteger e = new BigInteger(scanner.nextLine());

                // Verify if e is valid
                if (!isValidE(e, phi)) {
                    System.out.println("Error: Invalid value for e!");
                    continue;
                }

                // Get d from user
                System.out.print("Choose d (d * e ≡ 1 (mod phi(n))): ");
                BigInteger d = new BigInteger(scanner.nextLine());

                // Verify if d is valid
                if (!isValidD(d, e, phi)) {
                    System.out.println("Error: Invalid value for d! It must satisfy d * e ≡ 1 (mod phi(n))");
                    continue;
                }

                keys[0] = e;  // Public Key (e)
                keys[1] = d;  // Private Key (d)
                keys[2] = n;  // Modulus (n)
                break;

            } catch (NumberFormatException ex) {
                System.out.println("Error: Please enter valid numbers!");
            }
        }
    }

    // Encrypt message using public key (e, n)
    static BigInteger encrypt(BigInteger m, BigInteger e, BigInteger n) {
        return power(m, e, n);
    }

    // Decrypt message using private key (d, n)
    static BigInteger decrypt(BigInteger c, BigInteger d, BigInteger n) {
        return power(c, d, n);
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        BigInteger[] keys = new BigInteger[3]; // e, d, n

        try {
            // Key Generation with user input
            generateKeys(keys, scanner);

            System.out.println("\nPublic Key (e, n): (" + keys[0] + ", " + keys[2] + ")");
            System.out.println("Private Key (d, n): (" + keys[1] + ", " + keys[2] + ")");

            // Get message from user
            System.out.print("\nEnter message (numeric value): ");
            BigInteger M = new BigInteger(scanner.nextLine());

            // Check if message is less than n
            if (M.compareTo(keys[2]) >= 0) {
                System.out.println("Error: Message must be less than n = " + keys[2]);
                return;
            }

            System.out.println("Original Message: " + M);

            // Encrypt the message
            BigInteger C = encrypt(M, keys[0], keys[2]);
            System.out.println("Encrypted Message: " + C);

            // Decrypt the message
            BigInteger decrypted = decrypt(C, keys[1], keys[2]);
            System.out.println("Decrypted Message: " + decrypted);

        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        } finally {
            scanner.close();
        }
    }
}

File: rsa.py
def is_prime(n):
    if n <= 1:
        return False
    if n <= 3:
        return True
    if n % 2 == 0:
        return False
    i = 3
    while i * i <= n:
        if n % i == 0:
            return False
        i += 2
    return True

def gcd(a, b):
    while b:
        a, b = b, a % b
    return a

def mod_inverse(e, phi):
    # Extended Euclidean Algorithm
    t, new_t = 0, 1
    r, new_r = phi, e
    while new_r != 0:
        quotient = r // new_r
        t, new_t = new_t, t - quotient * new_t
        r, new_r = new_r, r - quotient * new_r
    if r > 1:
        return None  # No inverse exists
    if t < 0:
        t += phi
    return t

def mod_pow(base, exponent, modulus):
    result = 1
    base = base % modulus
    while exponent > 0:
        if exponent % 2 == 1:
            result = (result * base) % modulus
        exponent = exponent // 2
        base = (base * base) % modulus
    return result

def is_valid_e(e, phi):
    return 1 < e < phi and gcd(e, phi) == 1

def is_valid_d(d, e, phi):
    return (d * e) % phi == 1

def generate_keys():
    while True:
        try:
            p = int(input("Enter first prime number (p): "))
            q = int(input("Enter second prime number (q): "))

            if not is_prime(p) or not is_prime(q):
                print("Error: Both numbers must be prime!")
                continue

            n = p * q
            phi = (p - 1) * (q - 1)
            print("phi(n) =", phi)

            e = int(input("Choose e (1 < e < phi(n) and gcd(e, phi(n)) = 1): "))
            if not is_valid_e(e, phi):
                print("Error: Invalid value for e!")
                continue

            d = int(input("Choose d (d * e ≡ 1 (mod phi(n))): "))
            if not is_valid_d(d, e, phi):
                print("Error: Invalid value for d!")
                continue

            return e, d, n
        except ValueError:
            print("Error: Please enter valid integers.")

def encrypt(m, e, n):
    return mod_pow(m, e, n)

def decrypt(c, d, n):
    return mod_pow(c, d, n)

def main():
    e, d, n = generate_keys()

    print(f"\nPublic Key (e, n): ({e}, {n})")
    print(f"Private Key (d, n): ({d}, {n})")

    try:
        m = int(input("\nEnter message (numeric value): "))
        if m >= n:
            print(f"Error: Message must be less than n = {n}")
            return

        print("Original Message:", m)
        c = encrypt(m, e, n)
        print("Encrypted Message:", c)
        decrypted = decrypt(c, d, n)
        print("Decrypted Message:", decrypted)
    except ValueError:
        print("Error: Please enter a valid numeric message.")

if __name__ == "__main__":
    main()
